lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
+  for (FullApplySite ai : applies) {
+    paiUsers.push_back(ai.getInstruction());
   }
 
-  if (needsDestroys) {
-    // Compute the set of endpoints, which will be used to insert destroys of
-    // temporaries. This may fail if the frontier is located on a critical edge
-    // which we may not split (no CFG changes in SILCombine).
-    ValueLifetimeAnalysis vla(pai);
-    if (!vla.computeFrontier(partialApplyFrontier,
-                             ValueLifetimeAnalysis::DontModifyCFG))
-      return false;
-  }
+  // Also include all destroys in the liferange for the arguments.
+  // This is needed for later processing in tryDeleteDeadClosure: in case the
+  // pai gets dead after this optimization, tryDeleteDeadClosure relies on that
+  // we already copied the pai arguments to extend their lifetimes until the pai
+  // is finally destroyed.
+  collectDestroys(pai, paiUsers);
 
-  for (auto argWithIdx : argsToHandle) {
-    SILValue Arg = argWithIdx.first;
-    builder.setInsertionPoint(pai->getFunction()->begin()->begin());
-    // Create a new temporary at the beginning of a function.
-    SILDebugVariable dbgVar(/*Constant*/ true, argWithIdx.second);
-    auto *tmp = builder.createAllocStack(pai->getLoc(), Arg->getType(), dbgVar);
-    builder.setInsertionPoint(pai);
-    // Copy argument into this temporary.
-    builder.createCopyAddr(pai->getLoc(), Arg, tmp, IsTake_t::IsNotTake,
-                           IsInitialization_t::IsInitialization);
+  ValueLifetimeAnalysis vla(pai, paiUsers);
+  ValueLifetimeAnalysis::Frontier partialApplyFrontier;
 
-    tmpCopies.push_back(tmp);
-    argToTmpCopy.insert(std::make_pair(Arg, tmp));
+  // Computing the frontier may fail if the frontier is located on a critical
+  // edge which we may not split.
+  if (!vla.computeFrontier(partialApplyFrontier,
+                           ValueLifetimeAnalysis::DontModifyCFG)) {
+    return false;
   }
-  return true;
-}
 
-/// Emit dealloc_stack for all temporaries.
-void PartialApplyCombiner::deallocateTemporaries() {
-  // Insert dealloc_stack instructions at all function exit points.
-  for (SILBasicBlock &block : *pai->getFunction()) {
-    TermInst *term = block.getTerminator();
-    if (!term->isFunctionExiting())
-      continue;
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
+    SILDebugVariable dbgVar(/*Constant*/ true, argIdx);
+    builder.setInsertionPoint(pai);
 
-    for (auto copy : tmpCopies) {
-      builder.setInsertionPoint(term);
-      builder.createDeallocStack(pai->getLoc(), copy);
+    SILValue tmp = arg;
+    if (arg->getType().isObject()) {
+      tmp = builder.emitCopyValueOperation(pai->getLoc(), arg);
+    } else {
+      // Copy address-arguments into a stack-allocated temporary.
+      tmp = builder.createAllocStack(pai->getLoc(), arg->getType(), dbgVar);
```

Comment:
I am fine with this now, but I do think that at some point, we should incorporate code from SILGen in terms of ManagedValue/Scopes here. Would make it much easier to work with.